File: solutions/0784. Letter Case Permutation/0784.py
# Time complexity: O(2^n)
# Space complexity: O(2^n)
class Solution:
  def letterCasePermutation(self, s: str) -> List[str]:
    def backtrack(path: List[str], i: int) -> None:
      if i == len(s):
        ans.append(''.join(path))
        return
      
      if s[i].isalpha():
        backtrack(path + [s[i].lower()], i + 1)
        backtrack(path + [s[i].upper()], i + 1)
      else:
        backtrack(path + [s[i]], i + 1)

    ans = []
    backtrack([], 0)

    return ans      

File: solutions/0784. Letter Case Permutation/0784.cpp
// # Time complexity: O(2^n)
// Space complexity: O(2^n)
class Solution {
 public:
  vector<string> letterCasePermutation(string s) {
    vector<string>ans;
    backtrack(s, 0, ans);

    return ans;
  }
 private:
  void backtrack(string& s, int i, vector<string>& ans) {
    if (i == s.length()) {
      ans.push_back(s);
      return;
    }
    if (isdigit(s[i])) {
      backtrack(s, i + 1, ans);
      return;
    }

    s[i] = tolower(s[i]);
    backtrack(s, i + 1, ans);
    s[i] = toupper(s[i]);
    backtrack(s, i + 1, ans);
  }
};